pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -163,25 +191,25 @@ public synchronized void registerCumulativeAckConsumer(ConsumerImpl<?> consumer)
     public CompletableFuture<Void> commit() {
         return checkIfOpenOrCommitting().thenCompose((value) -> {
             CompletableFuture<Void> commitFuture = new CompletableFuture<>();
-            this.state = State.COMMITTING;
-            allOpComplete().whenComplete((v, e) -> {
-                if (e != null) {
-                    abort().whenComplete((vx, ex) -> commitFuture.completeExceptionally(e));
-                } else {
-                    tcClient.commitAsync(new TxnID(txnIdMostBits, txnIdLeastBits))
-                            .whenComplete((vx, ex) -> {
-                                if (ex != null) {
-                                    if (ex instanceof TransactionNotFoundException
-                                            || ex instanceof InvalidTxnStatusException) {
-                                        this.state = State.ERROR;
-                                    }
-                                    commitFuture.completeExceptionally(ex);
-                                } else {
-                                    this.state = State.COMMITTED;
-                                    commitFuture.complete(vx);
+            executedFuture.thenRun(() -> {
```

Comment:
Should check if the `executedFuture` is completed? If false, we should give feedback to the caller?

Reply:
* We should wait all the ops are completely.
* If we return an exception, users have to use a loop and try catch to commit the txn.